### PR TITLE
DOMA-1899 [WIP] push notifications transport attempt #2

### PR DESCRIPTION
**Pull request by akarjakin**
- push transport support (Firebase by now), would be extended in future to support Huawei
- support for attempt of delivery of the same push notification to multiple pushTokens available (registered by user) with error handling
- Ticket event handlers (by now supported triggering notifications by type in different cases), mainly in afterChange hook
- transport prioritisation mechanism for push notifications
- fallback mechanism for push notifications transport priorities
- code refactoring

Message processingMeta for fallback transport

image

image

### CODE REVIEW

File: apps/condo/domains/notification/tasks.js [hunk 1]
```diff
@@ -29,44 +31,34 @@ async function _sendMessageByAdapter (transport, adapter, messageContext) {
...
-    if (message.status !== MESSAGE_SENDING_STATUS && message.status !== MESSAGE_RESENDING_STATUS) {
-        return `already-${message.status}`
-    }
-
-    const baseAttrs = {
-        // TODO(pahaz): it's better to use server side fingerprint?!
-        dv: message.dv,
-        sender: message.sender,
-    }
+    console.log('deliveryMessage:', { messageId, message, transport })
```
Comment (akarjakin) [inline apps/condo/domains/notification/tasks.js:60]: WIP

File: apps/condo/domains/notification/adapters/firebaseAdapter.js [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+const isEmpty = require('lodash/isEmpty')
+const isNull = require('lodash/isNull')
+const get = require('lodash/get')
+const faker = require('faker')
+
+const {
+    PUSH_FAKE_TOKEN_SUCCESS,
+    PUSH_FAKE_TOKEN_FAIL,
+    FIREBASE_CONFIG_ENV,
+} = require('../constants/constants')
```
Comment (AntonAL) [inline apps/condo/domains/notification/adapters/firebaseAdapter.js:11]: Looks weird

File: apps/condo/domains/notification/adapters/firebaseAdapter.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+let FIREBASE_CONFIG = null
+
+console.error('process.env.GOOGLE_RECAPTCHA_CONFIG.length:', get(process.env, 'GOOGLE_RECAPTCHA_CONFIG.length', undefined))
+
+if (!process.env[FIREBASE_CONFIG_ENV]) throw new Error(EMPTY_CONFIG_ERROR)
+
+// JSON.parse would throw for a broken JSON, so we should take care of it here
+try {
+    FIREBASE_CONFIG = JSON.parse(process.env[FIREBASE_CONFIG_ENV])
```
Comment (pahaz) [inline apps/condo/domains/notification/adapters/firebaseAdapter.js:24]: it's better to use `conf. FIREBASE_CONFIG`

File: apps/condo/domains/notification/adapters/firebaseAdapter.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+const get = require('lodash/get')
+const faker = require('faker')
+
+const {
+    PUSH_FAKE_TOKEN_SUCCESS,
+    PUSH_FAKE_TOKEN_FAIL,
+    FIREBASE_CONFIG_ENV,
+} = require('../constants/constants')
+
+const EMPTY_CONFIG_ERROR = `Valid ${FIREBASE_CONFIG_ENV} should be provided within .helm (.env), and can be retrieved from https://console.firebase.google.com/project/__PROJECT_ID__/settings/serviceaccounts/adminsdk`
```
Comment (AntonAL) [inline apps/condo/domains/notification/adapters/firebaseAdapter.js:13]: Condo app does not makes an assumption of how environment variables are provided. Make an error message agnostic from helm.

Comment (AntonAL) [inline apps/condo/domains/notification/adapters/firebaseAdapter.js:13]: Could you please extract errors into `constant/errors`

Comment (akarjakin) [reply apps/condo/domains/notification/adapters/firebaseAdapter.js:18]: Native error message is useless here, I made this special error message for a reason.

File: apps/condo/domains/notification/adapters/firebaseAdapter.js (hunk 2 above)
Comment (akarjakin) [reply apps/condo/domains/notification/adapters/firebaseAdapter.js:14]: Can you please tell some details? I can't catch the idea.

File: apps/condo/domains/notification/adapters/firebaseAdapter.js (hunk 3 above)
Comment (akarjakin) [reply apps/condo/domains/notification/adapters/firebaseAdapter.js]: done

File: apps/condo/domains/notification/adapters/firebaseAdapter.js (hunk 4 above)
Comment (akarjakin) [reply apps/condo/domains/notification/adapters/firebaseAdapter.js:18]: done

**Review by sitozzz — APPROVED** [review]

**Review by AntonAL — APPROVED** [review]

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/notification/ticketTemplates.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
+const get = require('lodash/get')
```
Comment (pahaz) [inline apps/condo/domains/notification/ticketTemplates.js:1]: Please add TODO, to remove this file in a future or don’t forget to remove it

Comment (akarjakin) [reply apps/condo/domains/notification/ticketTemplates.js:1]: I'm sure Alex will remove it when his translations PR will be rebasing to master.

**Pull request merged**